### ISSUE TITLE
Request: error codes?

### ISSUE THREAD

**Issue by FGasper** [chansen/p5-http-tiny#106]
Hello,

My team has need for detecting connection errors, e.g.:
```
Could not connect to '$host:$port'
```

Right now we’re parsing that text string. Would it be feasible to consider some sort of reporting mechanism that gives back a parsable object?

**Comment by xdg** [comment]
Is the issue detecting that *specific* string?  Or are you needing to get host/port out of it?

We control that string, so it's stable.  I'm not inclined to add OO error reporting -- that's not very "Tiny".  But if you can better describe your team's need, I'll consider what might be possible and still "Tiny".

**Comment by FGasper** [comment]
We’re wanting to detect this specific error condition and to extract the host/port out of it.

Maybe something like:

```
my $err_hr = $http->error_details();
```

… with `$err_hr` being something like:
```
{
    type => 'connect_error',
    host => 'the.host.name',
    port => 80,
}
```
???

I personally think object errors wouldn’t necessarily violate “Tiny” if, e.g., you lazy-load them only when they happen.

**Comment by xdg** [comment]
I don't think that saves you much over this:

```
eval { $http->get( $url ) };
if ( my $e =$@ ) {
    my ($host, $port) = $e =~ /Could not connect to '([^:]+):(\d+)'/);
    if ($host) { ... }
}
```

Having to call a method, check `type` (yet another magic string comparison), and only then look for `host` and `port` seems nearly as verbose and error prone:

```
eval { $http->get( $url ) };
if ( my $e =$@ ) {
    my $err_hr = $http->error_details($e);
    if ($err_hr->{type} eq 'connect_error') { ... }
}
```

What about something like:
```
eval { $http->get( $url ) };
if ( my $e =$@ ) {
    if ( my ($host,$port) = $http->is_connect_error($e) ) { ... }
}
```

Or, this -- which keeps the API surface smaller:
```
eval { $http->get( $url ) };
if ( my $e =$@ ) {
    if ( my ($host,$port) = $e =~ $HTTP::Tiny::MATCH_CONNECT_ERROR ) { ... }
}
```

**Comment by FGasper** [comment]
`yet another magic string comparison`
^^ ISTM that checking for equality against a known (documented) string constant is a very different thing from a pattern match against an English-language phrase. The latter carries no guarantee of consistency across versions (someone might rephrase it to, e.g., “Failed to connect: $host:$port”), whereas a published, easily searchable constant is implicitly reliable.

Even if it’s not much less code or even slightly more, the fact of its being reliable across versions and maintainers would be a big win.

Also, I wasn’t thinking to make these exceptions necessarily, since HTTP::Tiny doesn’t die() on most errors. (We have our own wrapper logic that provides that behavior with our internal exception framework.) Basically this would be a “details” whenever the error code is 599.

**Comment by FGasper** [comment]
@xdg A more fleshed-out proof of concept: https://github.com/FGasper/p5-http-tiny/commit/24d56038feed993f07a40201e4dbc5449d32e30a

There would be need of a way for the handle to tell the $http object about the last error, and a simple accessor, but otherwise this is it: simple, lightweight, and reliable across versions, languages, etc.

**Comment by xdg** [comment]
Advanced error reporting is outside the scope of HTTP::Tiny.  Closing.
